ffmpeg/lpms_ffmpeg.c

```diff
@@ -9,13 +9,52 @@
 #include <libavutil/opt.h>
 #include <libavutil/pixdesc.h>
 
-#include <pthread.h>
-
 // Not great to appropriate internal API like this...
 const int lpms_ERR_INPUT_PIXFMT = FFERRTAG('I','N','P','X');
 const int lpms_ERR_FILTERS = FFERRTAG('F','L','T','R');
 const int lpms_ERR_PACKET_ONLY = FFERRTAG('P','K','O','N');
 const int lpms_ERR_OUTPUTS = FFERRTAG('O','U','T','P');
+const int lpms_ERR_DTS = FFERRTAG('-','D','T','S');
+
+//
+//  Notes on transcoder internals:
+//
+//  Transcoding follows the typical process of the FFmpeg API:
+//    read/demux/decode/filter/encode/mux/write
+//
+//  This is done over discrete segments. However, decode/filter/encoder are
+//  expensive to re-initialize for every segment. We work around this by
+//  persisting these components across segments.
+//
+//  The challenge with persistence is there is often internal data that is
+//  buffered, and there isn't an explicit API to flush or drain that data
+//  short of re-initializing the component. This is addressed for each component
+//  as follows:
+//
+//  Decoder: For audio, we pay the price of closing and re-opening the decoder.
+//           For video, we cache the first packet we read (input_ctx.first_pkt).
+//           The pts is set to a sentinel value and fed to the decoder. Once we
+//           receive a frame from the decoder with the pts set to the sentinel
+//           value, then we know the decoder has been fully flushed. Ignore any
+//           subsequent frames we receive with the sentiel pts.
+//
+//  Filter:  The challenge here is around fps filter adding and dropping frames.
+//           The fps filter expects a strictly monotonic input pts: frames with
+//           earlier timestamps get dropped, and frames with too-late timestamps
+//           will see a bunch of duplicated frames be generated to catch up with
```

Comment:
What happens if segment missed? Like, orchestrator receives segment 1 and then segment 10?

Reply:
Great question. This actually works correctly. The fps filter duplicates the last frame to make up for the "missing frames" if it detects insertion(s) are necessary between the current frame and the last frame in order to maintain a smooth rate of output. If the current and last frames are "far apart", this does result in outputting a zillion duplicates to fill the gap.

But since the last frame is a flush frame, the duplicates are dropped before hitting the encoder. This leads to the final output still being correct. We actually have a test that implicitly checks this behavior via alternating timestamps in https://github.com/livepeer/lpms/pull/179/commits/4ab522fd3346432dda5e4d752cdbba45aa113490 which encodes segments 1 and 3 after dropping segments 0 and 2. But I'll see about adding an explicit check for skipped segments.

This brings up another issue I had overlooked until now. Since the fps filter comes *before* the scale filter, all duplicated frames get rescaled, then dropped. This is additional computation that really should be avoided since it is a clear attack vector: send artificial timestamps that are years apart and pin down the CPU / GPU as the filtergraph churns through all the duplicated frames.

https://github.com/livepeer/lpms/pull/179/commits/4ab522fd3346432dda5e4d752cdbba45aa113490

An easy fix for this would be to rearrange the filter string such that the fps adjustment comes *after* the rescaling, but it also means doing unnecessary work if the fps is being adjusted downwards (eg, 60fps -> 30fps) which is a rather common case (and that's the reason for the filters being ordered the way they are right now). But given the choice between doing redundant work, and protecting against an easy DoS,  it'd be better to rearrange the filters to avoid DoS.

That being said, I think it might be good to spend a little more time figuring out how to support alternating timestamps without making the code even more terrible than it already is, because I suspect we'll be running into ordering issues often with retries etc. Supporting alternating timestamps should allow us to preserve the current filtergraph ordering while mitigating the DoS.

Note that we've always had an issue with the user requesting an extremely high fps output, but theoretically that could be addressed by short circuiting the transcode as described in https://github.com/livepeer/lpms/issues/129

Reply:
> it'd be better to rearrange the filters to avoid DoS.

I agree we should do this until we implement support alternating timestamps.

Reply:
Also, this one https://github.com/livepeer/go-livepeer/issues/1422 will be an issue with persistent filtergraph too?

Reply:
> Also, this one livepeer/go-livepeer#1422 will be an issue with persistent filtergraph too?

It shouldn't, since timestamps are carried as int64 everywhere, and most sane formats can accommodate rollovers (including mpegts and RTMP). Dunno what the issue is with that ticket.

Filters rearranged and skip test added in https://github.com/livepeer/lpms/commit/c5d18249a287a6ddb32c8d2630ea4b8c2507435e